### DISCUSSION TITLE
[RFC]: Reduce package size

### DISCUSSION THREAD

**Discussion by tay1orjones** [IBM/plex#452] [Ideas]
**Background**

The current package, `@ibm/plex` has recently increased beyond the threshold of GitHub's large file size limit of 100 MB, https://github.com/IBM/plex/issues/453. Due to this, the compressed package binary (used for yarn cache for instance) is now ~170 MB and too large to be included in the project without enabling LFS. Yarn cache does not support LFS. We need to provide a way for these projects to include the package.

Below are two options to consider to resolve this. 

**Option 1 is my recommendation** after consulting various maintainers.

## Option 1: Provide per-family packages

We would like to provide packages for each individual family within Plex in addition to the existing `@ibm/plex` package. These new packages would be named using hyphenated suffixes:

- `@ibm/plex`
- `@ibm/plex-sans`
- `@ibm/plex-sans-condensed`
- `@ibm/plex-serif`
- ...etc

These packages will help provide a more concise way to include only specific assets within a project instead of the entire font. Additionally, the individual packages will be smaller, facilitating use within projects that vendor dependencies or use Git for dependency caching.

With per-family packages, projects could opt to install `@ibm/plex` as before, or instead install only the relevant families their project uses.

```
yarn add @ibm/plex-sans @ibm/plex-serif @ibm/plex-mono
```

**Technical Deep Dive**

To facilitate per-family packages the repository needs to be reconfigured to house the multiple packages under the one repository, known as a [monorepo](https://monorepo.tools/).

Additionally, tooling must be supplied to facilitate the management of the monorepo. Recently Carbon looked into updating it's monorepo tooling and found solid candidates with using yarn workspaces supplanted by `nx` or `turborepo`. 

Further, the packages within the monorepo need to be versioned and published to npm. [Turborepo's docs](https://turborepo.org/docs/guides/monorepo-tools#versioning-publishing-and-changelog-generation) outline some good candidates:

> - [changesets/changesets](https://github.com/changesets/changesets) - 🦋 A way to manage your versioning and changelogs with a focus on monorepos
> - [microsoft/beachball](https://github.com/microsoft/beachball) - The Sunniest Semantic Version Bumper
> - [lerna/lerna](https://github.com/lerna/lerna) - A tool for managing JavaScript projects with multiple packages.
> - [lerna/lerna-changelog](https://github.com/lerna/lerna-changelog) - PR-based changelog generator with monorepo support
> - [intuit/auto](https://github.com/intuit/auto) - Generate releases based on semantic version labels on pull requests.

Questions

- What tooling makes the most sense?
- Do we need to include `complete`, or can we just include `split`?
  - A: We will begin by only including `split` as this should cover the primary use cases. We can expand what is included in each package by request.
- Would this require a breaking change?
  - A: No, assets within `@ibm/plex` would not change

## Option 2: Remove certain families or file types

The bulk of the file size weight is consolidated within just a few families:

```
252K	./IBM-Plex-Sans-Variable
712K	./IBM-Plex-Sans-Hebrew
828K	./IBM-Plex-Sans-Thai
868K	./IBM-Plex-Sans-Thai-Looped
1.5M	./IBM-Plex-Sans-Arabic
2.1M	./IBM-Plex-Sans-Devanagari
2.5M	./css
3.7M	./IBM-Plex-Sans-Condensed
4.3M	./IBM-Plex-Mono
4.9M	./IBM-Plex-Serif
5.5M	./IBM-Plex-Sans
8.5M	./scss
 64M	./IBM-Plex-Sans-KR
 99M	./IBM-Plex-Sans-JP
```

We could remove Japanese and/or Korean to reduce the filesize. These could then be published as their own packages. We would likely still need a monorepo set up to manage this, which then negates any work/cost savings. If we're going to publish a different package for JP or KR, why not do it for all families?

Removing certain file types is another approach, but it doesn't result in a large savings:

- Removing otf/ttf won’t really help much (they’re like 16k each)
- Removing woff would give us 23M in JP complete, 31M in JP split

In either case the experience using Plex assets through `@ibm/plex` would be fragmented. It no longer would be the source of all assets. Additionally this would require a breaking change version as we'd be removing assets, which would not be backwards compatible.

Questions

- Would this require a breaking change?
  - A: Yes, assets within `@ibm/plex` would change

**Comment by kamekku14** [comment]
Well... I'll go with **Option 1**, if that's the case.

**Comment by mjabbink** [comment]
@davelab6 any insights you can share on this issue?

**Comment by mjabbink** [comment]
@tay1orjones Option 1 of the two.

**Comment by BoldMonday** [comment]
@tay1orjones Separate repos for the different sub families has always been my preference.

**Comment by tay1orjones** [comment]
One thing that's changed since I opened this proposal is that Nx has taken over stewardship of lerna. This makes lerna a viable contender and probably the default choice for monorepo instrumentation as it's what both the `carbon` and `carbon-for-ibm-dotcom` monorepos use.

**Comment by tay1orjones** [comment]
There is a question as to the organization of packages. I think it makes sense to just follow the way they're organized and delivered from @BoldMonday 

* `@ibm/plex-mono`
* `@ibm/plex-sans`
* `@ibm/plex-sans-arabic`
* `@ibm/plex-sans-condensed`
* `@ibm/plex-sans-devanagari`
* `@ibm/plex-sans-hebrew`
* `@ibm/plex-sans-jp`
* `@ibm/plex-sans-kr`
* `@ibm/plex-sans-tc`
* `@ibm/plex-sans-thai`
* `@ibm/plex-sans-thai-looped`
* `@ibm/plex-sans-variable`
* `@ibm/plex-serif`

I think it's worthwhile to **_avoid_** putting jp/kr/tc/sc into a single `@ibm/plex-sans-cjk` package because this runs the risk of running into the assets being too large to publish to npm, https://github.com/IBM/plex/issues/561. I think individual packages for jp, kr, tc, etc would be a prudent choice. Following the delivery/folder structure has another benefit of making it easier to administer the repo as a monorepo since each folder will effectively be it's own npm package.

**Reply by tay1orjones** [reply]
This would mean that `@ibm/plex-sans` would contain latin/cyrillic, for instance.

**Reply by tay1orjones** [reply]
cc @mjabbink see above

**Reply by BoldMonday** [reply]
Following the current folder structure makes sense. That is also how fonts are listed in font menus, for example.

**Reply by BoldMonday** [reply]
And splitting KR/JP/SC/TC into individual packages makes perfectly sense as well.

**Reply by MY1L** [reply]
我路过的，有些问题想说/
(I'm a layman) but SC/TC/JP… have many common Chinese characters, wouldn't that be duplicating🤔

(I have read https://github.com/IBM/plex/issues/554#issuecomment-1907088506)

----

`@ibm/plex-sans-variable`
does this mean:
`@ibm/plex-sans-condensed-variable`
`@ibm/plex-sans-jp-variable`
`@ibm/plex-serif-variable`
……

----

`@ibm/plex-sans-thai`
`@ibm/plex-sans-thai-looped`
this reminds me the https://notofonts.github.io/#thai
`Noto Looped Thai`
`Noto Sans Thai Looped`

“Thai Looped” may belong to “Thai”?

**Comment by mjabbink** [comment]
This makes sense to me. My only concern is not having two sets of sans, one with CY/GK and one without. A few years back, we found that some teams didn’t need to include Cyrillic and Greek and needed the font files within a specific size limit. 

This means we have a sans `without` CY/GK (Latin) and a Sans version `with` CY/GK.

**Comment by mjabbink** [comment]
Something like below?
- @ibm/plex-sans
- @ibm/plex-sans-extended

**Reply by tay1orjones** [reply]
I'm not opposed, but it would be best if the source on box delivered from @BoldMonday matched this folder structure if we did this.

**Reply by BoldMonday** [reply]
I think this needs more consideration.
The name `sans-extended` implies a wider version of `sans`, similar to how `sans-condensed` is a narrower version.

This proposal also means that font naming inside of the fonts needs to be distinctive between fonts `with` CY/GK and `without` CY/GK. Because otherwise there will be interoperability problems when documents are being exchanged.

There are split font files currently for Sans where Latin, Cyrillic and Greek are already separated. Is that not sufficient?

**Reply by MY1L** [reply]
maybe like—
`@ibm/plex-sans`
`@ibm/plex-sans-Slim` (without CY/GK = specific **s**ize **lim**it)

**Comment by mjabbink** [comment]
You’re totally right @BoldMonday. I have not had “extended” in my brain for a long time, but yes, we can't use that name.


- `@ibm/plex-mono`
- `@ibm/plex-sans`
- `@ibm/plex-sans-arabic`
- `@ibm/plex-sans-condensed`
- `@ibm/plex-sans-devanagari`
- `@ibm/plex-sans-hebrew`
- `@ibm/plex-sans-jp`
- `@ibm/plex-sans-kr`
- `@ibm/plex-sans-tc`
- `@ibm/plex-sans-sc`
- `@ibm/plex-sans-thai`
- `@ibm/plex-sans-thai-looped`
- `@ibm/plex-sans-variable`
- `@ibm/plex-serif`


```

**Comment by mjabbink** [comment]
@tay1orjones Let’s go with the above.
